### PR TITLE
Add content for saving to disk and Azure blob storage

### LINKED ISSUE #5160: RP tutorial file upload sample: Add save to disk
https://docs.microsoft.com/aspnet/core/tutorials/razor-pages/uploading-files
/aspnetcore/tutorials/razor-pages/uploading-files.md

The RP tutorial file upload sample currently saves the content of the file to the dB. That's fine; however, two devs have asked about saving the contents to disk server-side. We could add ...

* Active code to the sample to show how to do it with notes on enabling the worker process to write to the location on disk.
* Inactive commented code that shows how to do it without getting deeply into write permissions instructions.

Either way, the code will be something along the lines of ...

```csharp
using (var fileStream = new FileStream(filePath, FileMode.Create))
{
    await formFile.CopyToAsync(fileStream);
}
```

@Rick-Anderson @scottaddie I can knock this out within a week or two.

cc/ @Randall_Faldon

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #5160 

[Internal Review Topic (Security considerations)](https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/uploading-files?branch=pr-en-us-5244#security-considerations)
[Save the file to disk](https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/uploading-files?branch=pr-en-us-5244#save-the-file-to-disk)
Save file to Azure Blob Storage: Immediately under the *Save the file to disk* section. :point_up:

Decided we should try a general security section with this update.

@blowdart Need ur wisdom here sir. :beer:

### CODE REVIEW

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 1]
```diff
@@ -20,6 +20,24 @@ The [Razor Pages Movie sample app](https://github.com/aspnet/Docs/tree/master/as
 
 In the steps below, you add a movie schedule file upload feature to the sample app. A movie schedule is represented by a `Schedule` class. The class includes two versions of the schedule. One version is provided to customers, `PublicSchedule`. The other version is used for company employees, `PrivateSchedule`. Each version is uploaded as a separate file. The tutorial demonstrates how to perform two file uploads from a page with a single POST to the server.
 
+## Security considerations
+
+Great caution must be taken when providing users with the ability to upload files to a server. Attackers may execute [denial of service](/windows-hardware/drivers/ifs/denial-of-service) and other attacks on a system. Some of the security steps that must be taken to reduce the likelihood that an attack will succeed are:
```
Comment (Rick-Anderson) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: ~Great~

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 2]
```diff
@@ -20,6 +20,24 @@ The [Razor Pages Movie sample app](https://github.com/aspnet/Docs/tree/master/as
...
+
+* Upload files to a dedicated file upload area on the system, which makes it easier to impose security measures on uploaded content. When permitting file uploads, make sure that execute permissions are disabled on the upload location.
+* Use a safe filename determined by the app, not from user input or the filename of the uploaded file.
+* Only allow a specific set of approved file extensions.
+* Make sure that any client-side checks are also made by the app. It's easy for attackers to circumvent client-side checks, so never trust checks made in JavaScript on the client.
```
Comment (Rick-Anderson) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: Verify  client-side checks are also performed on the server. Client side checks are easy to circumvent.

File: aspnetcore/tutorials/razor-pages/razor-pages-start/sample/RazorPagesMovie/Utilities/FileHelpers.cs [hunk 3]
```diff
@@ -57,6 +57,10 @@ public class FileHelpers
             {
                 modelState.AddModelError(formFile.Name, $"The {fieldDisplayName}file ({fileName}) is empty.");
             }
+            else if (formFile.Length > 1048576)
+            {
+                modelState.AddModelError(formFile.Name, $"The {fieldDisplayName}file ({fileName}) exceeds 1 MB.");
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/razor-pages-start/sample/RazorPagesMovie/Utilities/FileHelpers.cs]: Should there be a space between `{fieldDisplayName}` and `file`?

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 4]
```diff
@@ -20,6 +20,24 @@ The [Razor Pages Movie sample app](https://github.com/aspnet/Docs/tree/master/as
 
 In the steps below, you add a movie schedule file upload feature to the sample app. A movie schedule is represented by a `Schedule` class. The class includes two versions of the schedule. One version is provided to customers, `PublicSchedule`. The other version is used for company employees, `PrivateSchedule`. Each version is uploaded as a separate file. The tutorial demonstrates how to perform two file uploads from a page with a single POST to the server.
 
+## Security considerations
+
+Caution must be taken when providing users with the ability to upload files to a server. Attackers may execute [denial of service](/windows-hardware/drivers/ifs/denial-of-service) and other attacks on a system. Some of the security steps that must be taken to reduce the likelihood that an attack will succeed are:
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: ~~Some of the security steps that must be taken to reduce the likelihood that an attack will succeed are:~~
Some security steps to reduce the likelihood that an attack succeeds are:

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 5]
```diff
@@ -20,6 +20,24 @@ The [Razor Pages Movie sample app](https://github.com/aspnet/Docs/tree/master/as
...
+
+* Upload files to a dedicated file upload area on the system, which makes it easier to impose security measures on uploaded content. When permitting file uploads, make sure that execute permissions are disabled on the upload location.
+* Use a safe filename determined by the app, not from user input or the filename of the uploaded file.
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: Should "filename" be two words?

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 6]
```diff
@@ -20,6 +20,24 @@ The [Razor Pages Movie sample app](https://github.com/aspnet/Docs/tree/master/as
...
+* Only allow a specific set of approved file extensions.
+* Verify client-side checks are also performed on the server. Client side checks are easy to circumvent.
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: Avoid using the word "also".

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 7]
```diff
@@ -20,6 +20,24 @@ The [Razor Pages Movie sample app](https://github.com/aspnet/Docs/tree/master/as
...
+> [!WARNING]
+> Uploading malicious code to a system is frequently the first step to executing code that can:
+> * Completely takeover a system.
+> * Overload a system with the result that the system completely fails.
+> * Compromise user or system data.
+> * Apply graffiti to a public interface.
+
 ## Add a FileUpload class
 
 Below, you create a Razor page to handle a pair of file uploads. Add a `FileUpload` class, which is bound to the page to obtain the schedule data. Right click the *Models* folder. Select **Add** > **Class**. Name the class **FileUpload** and add the following properties:
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: Razor page --> Razor Page

File: aspnetcore/tutorials/razor-pages/razor-pages-start/sample/RazorPagesMovie/Utilities/FileHelpers.cs (hunk 3 above)
Comment (guardrex) [reply aspnetcore/tutorials/razor-pages/razor-pages-start/sample/RazorPagesMovie/Utilities/FileHelpers.cs]: The space is added as part of the `fieldDisplayName`; so if the field is empty, an extra space isn't present in the string.

File: aspnetcore/tutorials/razor-pages/uploading-files.md (hunk 4 above)
Comment (guardrex) [reply aspnetcore/tutorials/razor-pages/uploading-files.md]: Let's try this one on the next round:

> Some security steps that reduce the likelihood of a successful attack are:

**Comment by guardrex** [comment]
Thanks @scottaddie ... How does it look with those changes?

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 8]
```diff
@@ -18,11 +18,29 @@ In this section, uploading files with a Razor Page is demonstrated.
...
 
-In the steps below, you add a movie schedule file upload feature to the sample app. A movie schedule is represented by a `Schedule` class. The class includes two versions of the schedule. One version is provided to customers, `PublicSchedule`. The other version is used for company employees, `PrivateSchedule`. Each version is uploaded as a separate file. The tutorial demonstrates how to perform two file uploads from a page with a single POST to the server.
+In the steps below, a movie schedule file upload feature is added to the sample app. A movie schedule is represented by a `Schedule` class. The class includes two versions of the schedule. One version is provided to customers, `PublicSchedule`. The other version is used for company employees, `PrivateSchedule`. Each version is uploaded as a separate file. The tutorial demonstrates how to perform two file uploads from a page with a single POST to the server.
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: ~~In the steps below~~
In the following steps

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 9]
```diff
@@ -18,11 +18,29 @@ In this section, uploading files with a Razor Page is demonstrated.
...
+
+## Security considerations
+
+Caution must be taken when providing users with the ability to upload files to a server. Attackers may execute [denial of service](/windows-hardware/drivers/ifs/denial-of-service) and other attacks on a system. Some security steps that reduce the likelihood of a successful attack are:
+
+* Upload files to a dedicated file upload area on the system, which makes it easier to impose security measures on uploaded content. When permitting file uploads, make sure that execute permissions are disabled on the upload location.
+* Use a safe file name determined by the app, not from user input or the file name of the uploaded file.
+* Only allow a specific set of approved file extensions.
+* Verify client-side checks are performed on the server. Client side checks are easy to circumvent.
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: Should the "Client side" reference starting this sentence be hyphenated?

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 10]
```diff
@@ -18,11 +18,29 @@ In this section, uploading files with a Razor Page is demonstrated.
...
+> Uploading malicious code to a system is frequently the first step to executing code that can:
+> * Completely takeover a system.
+> * Overload a system with the result that the system completely fails.
+> * Compromise user or system data.
+> * Apply graffiti to a public interface.
 
 ## Add a FileUpload class
 
-Below, you create a Razor page to handle a pair of file uploads. Add a `FileUpload` class, which is bound to the page to obtain the schedule data. Right click the *Models* folder. Select **Add** > **Class**. Name the class **FileUpload** and add the following properties:
+Below, create a Razor Page to handle a pair of file uploads. Add a `FileUpload` class, which is bound to the page to obtain the schedule data. Right click the *Models* folder. Select **Add** > **Class**. Name the class **FileUpload** and add the following properties:
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: Avoid using the word "below", per Rick.

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 11]
```diff
@@ -139,7 +174,7 @@ The user can click the **Delete** link from there to reach the delete confirmati
 
 For troubleshooting information with `IFormFile` uploading, see the [File uploads in ASP.NET Core: Troubleshooting](xref:mvc/models/file-uploads#troubleshooting).
 
-Thanks for completing this introduction to Razor Pages. We appreciate any comments you leave. [Getting started with MVC and EF Core](xref:data/ef-mvc/intro) is an excellent follow up to this tutorial.
+Thanks for completing this introduction to Razor Pages. We appreciate any comments left. [Getting started with MVC and EF Core](xref:data/ef-mvc/intro) is an excellent follow up to this tutorial.
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: "comments left" --> "feedback"

File: aspnetcore/tutorials/razor-pages/uploading-files.md [hunk 12]
```diff
@@ -34,6 +52,23 @@ To avoid code duplication for processing uploaded schedule files, add a static h
...
+{
+    await formFile.CopyToAsync(fileStream);
+}
+```
+
+The worker process must have write permissions to the location specified by `filePath`.
+
+### Save the file to Azure Blob Storage
+
+To upload file content to Azure Blob Storage, see [Get started with Azure Blob storage using .NET](/azure/storage/blobs/storage-dotnet-how-to-use-blobs). The topic demonstrates how to use [UploadFromStream](/dotnet/api/microsoft.windowsazure.storage.file.cloudfile.uploadfromstreamasync) to save a [FileStream](/dotnet/api/system.io.filestream) to blob storage.
```
Comment (scottaddie) [inline aspnetcore/tutorials/razor-pages/uploading-files.md]: In the link text, change "storage" to "Storage".

**Comment by guardrex** [comment]
@scottaddie Updates made :+1:

**Comment by scottaddie** [comment]
@guardrex Looks good to me. Let's have @blowdart review before merging.

**Comment by Rick-Anderson** [comment]
@HaoK  would you be able to sign of on this smallish change?

**Comment by guardrex** [comment]
@HaoK There's a new security bit in the topic now: https://review.docs.microsoft.com/en-us/aspnet/core/tutorials/razor-pages/uploading-files?branch=pr-en-us-5244#security-considerations

**Comment by HaoK** [comment]
Looks fine to me, @blowdart should review when he gets a chance

**Comment by guardrex** [comment]
ping @blowdart

**Comment by blowdart** [comment]
LGTM

**Pull request merged**